modules/src/timestamp.rs

```diff
@@ -0,0 +1,179 @@
+use serde_derive::{Deserialize, Serialize};
+use std::convert::TryInto;
+use std::fmt::Display;
+use std::num::{ParseIntError, TryFromIntError};
+use std::str::FromStr;
+use thiserror::Error;
+
+use chrono::{offset::Utc, DateTime, TimeZone};
+
+/// A newtype wrapper over `Option<DateTime<Utc>>` to keep track of
+/// IBC packet timeout.
+///
+/// We use an explicit `Option` type to distinguish this when converting between
+/// a `u64` value and a raw timestamp. In protocol buffer, the timestamp is
+/// represented as a `u64` Unix timestamp in nanoseconds, with 0 representing the absence
+/// of timestamp.
+#[derive(PartialEq, Eq, Copy, Clone, Debug, Deserialize, Serialize, Hash)]
+pub struct Timestamp {
+    time: Option<DateTime<Utc>>,
```

Comment:
Rather than wrap an `Option<DateTime<Utc>>`, how about returning an `Option<Timestamp>` from `from_nanoseconds`? ie.

```rust
pub fn from_nanoseconds(nanoseconds: u64) -> Result<Option<Timestamp>, TryFromIntError> {
``

Reply:
I did considered that. However that cause issue because we can't derive or implement additional traits on `Option`. So the other alternative is to define a separate `OptionalTimestamp` and `Timestamp` newtypes. But looking around the usage, I think all current use of timestamp need to consider the null value.

Reply:
Alright, makes sense!